src/Sarif/Core/Stack.cs

```diff
@@ -0,0 +1,166 @@
+            Queue<Exception> exceptions;
+
+            stacks = new List<Stack>();
+            exceptions = new Queue<Exception>(new Exception[] { exception });
+
+            while (exceptions.Count > 0)
+            {
+                Stack stack;
+                Exception current;
+                
+                current = exceptions.Dequeue();
+
+                if (current.InnerException != null)
+                {
+                    exceptions.Enqueue(current.InnerException);
+                }
+
+                stack = Create(current.StackTrace);
+
+                stack.Message = current.Message;
+
+                stacks.Add(stack);
+            }
+
+            return stacks;
+        }
+
+        public static Stack Create(string stackTrace)
+        {
+            Stack stack = new Stack();
+            
+            if (string.IsNullOrEmpty(stackTrace))
+            {
+                return stack;
+            }
+
+            stack.Frames = new List<StackFrame>();
+
+            foreach (string line in stackTrace.Split('\n'))
+            {
+                // at Type.Method() in File.cs : line X\r
+                string current = line;
+
+                StackFrame stackFrame;
+                stackFrame = new StackFrame();
+
+                // at Type.Method() in File.cs: line X
+                current = line.Trim('\r');
+
+                // Type.Method() in File.cs: line X
+                current = current.Substring(AT.Length);
+                stackFrame.FullyQualifiedLogicalName = current.Split(')')[0] + ")";
+
+                // in File.cs: line X
+                current = current.Substring(stackFrame.FullyQualifiedLogicalName.Length);
+
+                if (current.Length > 0)
+                {
+                    int lineNumber;
+                    string fileName;
+
+                    // File.cs: line X
+                    current = current.Substring(IN.Length);
+
+                    // :line X
+                    fileName = current.Split(' ')[0];
+                    fileName = fileName.Substring(0, fileName.Length - LINE.Length);
+
+                    // X
+                    current = current.Substring(fileName.Length + LINE.Length);
+                    lineNumber = int.Parse(current);
+
+                    stackFrame.Location = new PhysicalLocation()
+                    {
+                        Uri = new Uri(fileName),
+                        Region = new Region()
+                        {
+                           StartLine = lineNumber
+                        }
+                    };
+                }
+                stack.Frames.Add(stackFrame);
+            }
+
+            return stack;
+        }
+
+        public override string ToString()
+        {
+            return ToString(StackFormat.Default);
+        }
+
+        public string ToString(StackFormat stackFormat)
+        {           
+            if (this.Frames == null) { return null; }
```

Comment:
Even string.Empty is discouraged (but still much better than null). See notes to inheritors: [https://msdn.microsoft.com/en-us/library/system.object.tostring%28v=vs.110%29.aspx.](https://msdn.microsoft.com/en-us/library/system.object.tostring%28v=vs.110%29.aspx.) 

---
In reply to: [60651545](https://github.com/Microsoft/sarif-sdk/pull/205#discussion_r60651545) [](ancestors = 60651545)